test/analysis_options.yaml

```diff
@@ -0,0 +1,19 @@
+analyzer:
```

Comment:
Não seria viável dar include nas regras do lint do diretório root e aqui apenas desabilitar as que não fizerem sentido para os testes?

Reply:
Minha sugestão é isolar os ambientes, permitindo aplicar regras de Lint adequado a necessidade de cada ambiente. Criar uma estrutura de hierarquia com estas pode gerar confusão e efeitos colaterais, que poderá ser evitado sendo obrigado a declarar os Lints em comum em ambos os ambiente.

Reply:
Até mesmo extender algumas das regras já prontas nos packages de lint como o `package:flutter_lints/flutter.yaml` teriamos esses problemas? estou pensando mais na consistência de código, além dos relacionados ao `const` e `invalid_null_aware_operator` teriam mais outras que precisa ser "relaxada"?

Reply:
O meu ponto é que adotar a politica de gerenciar regra de excessão é complicado ao médio e longo prazo. Prefiro o caminho onde está claro quais são as regras de Linter que está sendo aplicado para aquele ambiente, e não ficar trabalhando com excessão.

Reply:
Dando meus dois cents aqui, tendo a preferir a linha que o Pedro está propondo. Ter arquivos de config totalmente independentes por ambiente costuma levar a "falta de sincronia" e, em geral, um dos "ambientes" costuma estar up-to-date e os demais ficam defasados.

Mas assim.... alguém consegue [ELI5](https://www.dictionary.com/e/slang/eli5/) porque precisa dessa diferença? Não entendi ainda....

Disclaimer: Não manjo nada de desenvolvimento mobile e/ou flutter.

Reply:
## Definições

**Código de produção** é código que será analisado, parseado, compilado e publicado na loja. É em resumo o código que o usuário final vai consumir;

**Código de teste** é o código de apoio para validar as regras do código de produção, evitando regressão de comportamento e lógica de negócio;

**Linter** para este ambiente será considerado um código, uma regra de instruções que controla o comportamento do projeto.

## Afirmação

"Código de teste é diferente de código de produção, desta maneira pode possuir dependências e regras exclusivas para este ambiente".

###  Concordo com esta afirmação

Se for aceito que o código de teste é diferente do código de produção, então podemos afirmar que as regras aplicadas a uma pode não ser aplicado ao outro.

Neste ponto não importa ter Linter compartilhado por ambos, no desejo de evitar duplicidade. Neste momento do projeto podemos ter um código que pode ser utilizado por dois consumidores diferente, mas no futuro próximo próximo o motivo de alterar este código será diferente para cada cenário.

Neste momento é obrigatório utilizar `const` na construção de um objeto imutável no código de produção. Só que este é um comportamento não desejado no ambiente de teste, porque podemos explicitamente  querer comparar objetos imutáveis.

```language:dart
  User() == User()

```


### Discordo desta afirmação

Código é código, teste é código, portanto todo código no projeto precisa seguir a mesma regra de comportamento no projeto. Uma das maneiras que temos que definir regra é através do Linter, e portanto o mesmo Linter será aplicado todos os códigos dentro do projeto.

Reply:
Entendi, obrigado pela resposta @shonorio!

Da minha experiência, com as tecnologias e ambientes nos quais eu trabalhei, tendo a concordar com a afirmação de que `código é código, teste é código, portanto todo código no projeto preicsa seguir a mesma regra de comportamento no projeto`.

Mas sei que existem nuances entre linguagens e plataformas.

Nesse sentido, e considerando o que você colocou, eu diria que devemos optar por:
1. O mesmo setting de linter para todo mundo; ou
2. Um setting de linter para cada ambiente, totalmente independente.

Reply:
Durante os meus estudos de patterns, encontrei uma citação do [Robert C. Martin (Uncle Bob)](https://en.wikipedia.org/wiki/Robert_C._Martin) no livro Working Effectively with Legacy Code.

> Test code doesn't have to live up to the same standards as production code. In general, I don't mind breaking encapsulation by making variables public if it makes it easier to write tests. However, test code  should be clean. I should be easy to understand and change.

Se olharmos os códigos que estão subindo, encontraremos vários códigos com **ignore_for_** (desativando o linter), o que é um anti-pattern. O projeto está configurado para um mundo tão idealizado e fora da realidade atual do projeto nos obrigando a correr para um outro anti-pattern. Daqui a algum tempo quem vai revisitar o código para remover este anti-pattern?

Pelas referências acadêmicas sobre testes e, em menor grau, minha experiência profissional, recomendo que o ambiente de teste tenha um Linter, mas seja um linter apenas deste ambiente.

Reply:
Me pareceu uma boa argumentação e concordo em seguir desta forma.

Reply:
@shonorio você tem algum comentário sobre essa sugestão do @pedrox-hs ? De colocarmos mais algumas regras? (é isso Pedro?)

Reply:
Estas são as regras que utilizo para o ambiente de teste nos meus projetos. De todas as regras do Flutter Lint, a minha experiencia é que poucas vão uteis para o ambiente de teste

Reply:
> Estas são as regras que utilizo para o ambiente de teste nos meus projetos. De todas as regras do Flutter Lint, a minha experiencia é que poucas vão uteis para o ambiente de teste

eu tenho uma opinião contrária, pois algumas previnem algum erro de digitação (como `avoid_single_cascade_in_expression_statements` e `avoid_returning_null_for_void`) e outras melhoram a questão de padrão de estilo de código, evitando que fique muito diferente do código produtivo.
Resumindo, acredito que somente as regras que realmente "atrapalham" que não deveriam não serem incluidas.

Reply:
@pedrox-hs você tem uma proposta mais concreta de opções aqui?

Reply:
@diraol Honestamente eu sou mais a favor de manter uma hierarquia de regras, mesmo que para isso seja necessário desabilitar algumas regras, assim garantiria uma certa consistencia e pela nossa experiência no projeto, nunca tive problemas com as regras a ponto de necessitar atualizá-las constantemente.
Até seguindo o recomendado pela comunidade do Dart, isso somente seria prejudicial se eles fossem constantemente indecisos com as regras e formos atualizando a lib deliberadamente.
Mas se o caminho for realmente definir todas as regras, eu diria para copiarmos esse arquivo para o projeto: https://github.com/dart-lang/lints/blob/main/lib/recommended.yaml.

Reply:
Oi gente, voltando aqui.

Conversamos bastante essa semana última sobre isso e decidimos por uma abordagem "extensiva".

Para tanto, criei o PR #134 com o objetivo de implementar tal abordagem já levando o que consegui "pescar" de sugestões de "exclusões" aqui desse PR.

@shonorio e @pedrox-hs vocês podem olhar lá no #134, por favor?

Ah, e eu fiz um teste que, em teoria, deveria aplicar os testes em toda nossa codebase e ele passou:
https://github.com/institutoazmina/penhas-app/actions/runs/6336218079/job/17208882019?pr=134

Para esse teste eu mudei nosso Workflow para remover a restrição de aplicar as regras só nos arquivos modificados. Pelo que eu entendi o formatter foi aplicado em 703 arquivos e nenhum foi modificado.

Se puderem conferir se faz sentido eu agradeço!